web/src/lib/actions/input.ts

```diff
@@ -0,0 +1,229 @@
+  // macOS keybindings
+  mac: [
+    { key: 'q', meta: true }, // Quit
+    { key: 'w', meta: true }, // Close window
+    { key: 'h', meta: true }, // Hide
+    { key: 'm', meta: true }, // Minimize
+    { key: 'Tab', meta: true }, // App switcher
+    { key: ' ', meta: true }, // Spotlight
+    { key: 'F3', ctrl: true }, // Mission Control
+  ],
+  // Windows keybindings
+  win: [
+    { key: 'F4', alt: true }, // Close window
+    { key: 'Delete', ctrl: true, alt: true },
+    { key: 'Meta' }, // Start menu
+    { key: 'l', meta: true }, // Lock
+    { key: 'd', meta: true }, // Desktop
+    { key: 'Tab', meta: true }, // Task switcher
+  ],
+  // Linux keybindings
+  linux: [{ key: 'F4', alt: true }, { key: 'Delete', ctrl: true, alt: true }, { key: 'Meta' }],
+  // Browser-specific keybindings (cross-platform)
+  browser: [
+    { key: 't', ctrl: true },
+    { key: 'w', ctrl: true },
+    { key: 'n', ctrl: true },
+    { key: 'n', ctrl: true, shift: true },
+    { key: 't', ctrl: true, shift: true },
+    { key: 'p', ctrl: true, shift: true },
+  ],
+};
+const ALL_RESERVED_SHORTCUTS = [
+  ...RESERVED_SHORTCUTS.mac,
+  ...RESERVED_SHORTCUTS.win,
+  ...RESERVED_SHORTCUTS.linux,
+  ...RESERVED_SHORTCUTS.browser,
+];
+
+const attachmentFactory = (listenerFactory: KeyDownListenerFactory) => (element: KeyTargets) => {
+  const listener = listenerFactory(element);
+  element.addEventListener('keydown', listener as EventListener);
+  return () => {
+    element.removeEventListener('keydown', listener as EventListener);
+  };
+};
+
+function isKeyboardEvent(event: Event): event is KeyboardEvent {
+  return 'key' in event;
+}
+
+/** Determines whether an event should be ignored. The event will be ignored if:
+ *  - The element dispatching the event is not the same as the element which the event listener is attached to
+ *  - The element dispatching the event is an input field
+ */
+export const shouldIgnoreEvent = (event: KeyboardEvent | ClipboardEvent): boolean => {
+  if (event.target === event.currentTarget) {
+    return false;
+  }
+  const type = (event.target as HTMLInputElement).type;
+  return ['textarea', 'text', 'date', 'datetime-local', 'email', 'password'].includes(type);
+};
+
+function isPromise<T = unknown>(obj: unknown): obj is Promise<T> {
+  return (
+    obj !== null &&
+    (typeof obj === 'object' || typeof obj === 'function') &&
+    typeof (obj as { then?: unknown }).then === 'function'
+  );
+}
+
+function checkForReservedShortcuts(
+  shortcuts: KeyCombo[],
+  handler: (event: KeyboardEvent) => void,
+): (event: KeyboardEvent) => void {
+  const formatCombo = (combo: KeyCombo) => {
+    const parts = [];
+    if (combo.ctrl) {
+      parts.push('Ctrl');
+    }
+    if (combo.alt) {
+      parts.push('Alt');
+    }
+    if (combo.shift) {
+      parts.push('Shift');
+    }
+    if (combo.meta) {
+      parts.push('Meta');
+    }
+    parts.push(combo.key);
+    return parts.join('+');
+  };
+
```

Comment:
I would personally move that outside the `checkForReservedShortcuts`, it seems like a rather generic function to me